_posts/2017-04-18-release-0.3p2.md

```diff
@@ -0,0 +1,57 @@
+---
+layout: post
+title: "Eve Alpha v0.3 preview 2"
+author: "Corey Montella"
+tags: []
+---
+
+_(Eve is a new programming language, and this is our development blog. If you’re new to Eve, [start here](http://play.witheve.com))_
+
+Now that we've got the v0.3 preview out the door, we're ramping up to a more regular release schedule. In that spirit, today we're releasing the second preview of Eve v0.3, which largely improves built in watchers, but contains some other goodies as well. 
+
+### What's new?
+
+Let's take a look at some of the bigger improvements.
+
+#### Embedding Eve Elements in Your App
+
+You can now embed Eve rendered elements into your non-Eve application. This is accomplished with the `addExternalRoot(tag:string, element:HTMLElement)` function, which registers an arbitrary root element to serve as parents to Eve-rendered elements. It works like this:
+
+```javascript
+import {Program} from "witheve";
+
+let program = new Program("my program");
+let htmlWatcher = program.attach("html");
+
+let eveWrapper = document.querySelector("#eve-wrapper");
+htmlWatcher.addExternalRoot("my-root", eveWrapper);
+```
+
+This creates a record tagged "my-root" in Eve, which you can treat as any other Eve element.
+
+```javascript
+program.bind("render into #eve-wrapper", ({find, record}) => {
+  let my_root = find("my-root");
+  return [
+    my_root.add("children", [
+      record("html/element", {tagname: "div", text: "hello, world!"})
+    ])
+  ];
+})
+```
+
+This scenario works well if you have an already existing application, and you'd like to use Eve to render certain elements.
+
+#### Watchers
+
+We've added and streamlined various watchers that interact with the browser. We're now at the point where v0.3 is caught up to v0.2.x in terms of the available watchers. Additionally, we've added:
+
+- `button` attribute to all mouse events (values `left`, `right`, `middle`, `4`, and `5`.
+- `double-click`, `mouse-down`, and `mouse-up` events.
+- `html/listener/context-menu` tag. If an element with this tag is right clicked, the default context menu will not be shown, allowing you to render your own. You can grab page-relative mouse position off of any of the mouse events as `page-x` and `page-y`.
+- `key-press` event, with support for printable characters and the following control keys: `tab`, `enter`, `shift`, `control`, `alt`, `escape`, `left`, `up`, `right`, `down`, and `meta`.
+- `focus` and `blur` events.
+
+#### Standard Library
+
+Add [`codepoint_length(text:string)`](https://github.com/witheve/Eve/pull/823/commits/58af002ec4c3807b0a91e3ee148e6f7a5359387a) function.
```

Comment:
We should also mention that we added support for creating expressions with multiple outputs (e.g. `math.range(start, stop)` is now available. Also that's `string.codepoint_length`.